crates/sui-storage/src/event_store/sql.rs

```diff
@@ -353,51 +364,70 @@ const QUERY_BY_RECIPIENT: &str = "SELECT * FROM events WHERE timestamp >= ? AND
 const QUERY_BY_OBJECT_ID: &str = "SELECT * FROM events WHERE timestamp >= ? AND \
     timestamp < ? AND object_id = ? ORDER BY timestamp ASC LIMIT ?";
 
+/// Maximum number of rows to insert at once as a batch.  SQLite has 64k limit in binding values.
+const MAX_INSERT_BATCH: usize = 1000;
+
 #[async_trait]
 impl EventStore for SqlEventStore {
     #[instrument(level = "debug", skip_all, err)]
     async fn add_events(&self, events: &[EventEnvelope]) -> Result<u64, SuiError> {
-        // TODO: submit writes in one transaction/batch so it won't just fail in the middle
         let mut cur_seq = self.seq_num.load(Ordering::Acquire);
         let initial_seq = cur_seq;
-        let mut rows_affected: u64 = 0;
-
-        // TODO: benchmark
-        // TODO: use techniques in https://docs.rs/sqlx-core/0.5.13/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values
-        // to execute all inserts in a single statement?
-        // TODO: See https://kerkour.com/high-performance-rust-with-sqlite
-        for event in events {
-            // Skip events that have a lower sequence number... which must be same or increasing
-            if event.seq_num < cur_seq {
-                debug!(tx_digest =? event.tx_digest, seq_num = event.seq_num, cur_seq, "Skipping event with lower sequence number than current");
+        let mut rows_affected = 0;
+
+        if events.is_empty() {
+            return Ok(0);
+        }
+
+        let mut start_index = 0;
+        let mut end_index = 0;
+
+        // Insert event in batches, skipping over events that have lower or decreasing sequence numbers
+        while end_index < events.len() {
+            // Skip events that have a lower sequence number. They've been seen before.
+            if events[start_index].seq_num < cur_seq {
+                start_index += 1;
+                end_index += 1;
```

Comment:
I think we do not need two while loops here. The first one (this one) is to fast forward to the first event that has `seq-num` larger than `cur_seq`. In the second loop we keep examining it. However, assuming the `events` we pass in is sorted by seq_num, we only need the first loop. Note the `events` have to be ordered chronically, otherwise the way we increase `cur_seq` every time we see a larger value is incorrect

Reply:
The reason why I added the second loop was because @sadhansood pointed out that we should check if subsequent events are still ordered.  Also the first loop condition that compares to cur_seq is just for events that are below cur_seq, so the logic is fine and that's why cur_seq is not updated.

Reply:
why would subsequent events not be in ordered though? If there's such a possibility we should fix it. For example, the events passed in somehow are in order: `[1, 1, 3, 3, 2, 2]`. When we use this logic, `cur_seq` will be updated to `3` and `2`s will be ignored which cause FN misses events.

Reply:
You're right that if we have such a scenario then it's a bug and we should fix it.  There shouldn't really be such a scenario, other than batch system restarting after an error and replaying older events, and thus this is intended to skip such old events.  It is true that right now within a batch this wouldn't happen, because add_events is only called for one TX at a time, but I'm intending to refactor the ingestion to batch multiple TXes and speed it up.